relay-server/src/actors/upstream.rs

```diff
@@ -81,6 +81,28 @@ impl UpstreamRequestError {
             _ => false,
         }
     }
+
+    fn is_permanent_rejection(&self) -> bool {
+        if let Self::ResponseError(status_code, response) = self {
+            // server sent a client error response ( trying to reauthenticate is unlikely to
+            // change the response so don't bother trying in the future)
+            if status_code.is_client_error() {
```

Comment:
If we need to defend against this logic changing 2 lines down, perhaps it makes more sense to do an exhaustive match on the StatusCode enum?

Reply:
Well the first condition catches all 4xx responses the second is a 403 with additional checks on the body so I don't know how I could write a match that would work (given that the conditions overlap). 

It is not really about being exhaustive in matching all variants, it is more about checking two different functional requirements one of which happens to include (at this time) the other one.

Reply:
This is incorrect. The spec says that a permanent rejection must fulfill all of the below criteria:

- Status code `403`
- JSON response body
- Field `"relay"` set to `"stop"`

This means only the below check should return true, but this block here must go. 

FWIW, we will change the authentication endpoint to emit explicit `"relay": "stop"` when we want Relay to back off.